src/cudamatrix/cu-kernels.cu

```diff
@@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
   }
 }
 
+template<typename Real>
+__global__
+static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
+                                  Real* obfj_terms, MatrixDim objf_dim, 
+                                  Real* obfj_derivs, MatrixDim derivs_dim) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
+  int32_cuda scores_index = i + j * scores_dim.stride;
+  Real K = 1.0 / (scores_dim.rows - 2.0);
+  Real L = scores[scores_index];
+  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
```

Comment:
I don't think you are really thinking about this in the spirit of
backprop.  The general principle is that you go forward computing the
objective function, and then you do a process that is roughly the
mirror-image of the forward process to backprop the derivatives through the
computation.

What I described was getting the derivatives of the objective function
w.r.t. the matrix of scores.  After that you just have to do the reverse of
the forward operations to get the derivatives w.r.t. S and the matrix of
xvectors.

Dan


On Sat, Feb 13, 2016 at 4:51 PM, david-ryan-snyder <notifications@github.com
> wrote:

> In src/cudamatrix/cu-kernels.cu
> <https://github.com/danpovey/kaldi/pull/5#discussion_r52834114>:
>
> > @@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
> >    }
> >  }
> >
> > +template<typename Real>
> > +__global__
> > +static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
> > +                                  Real* obfj_terms, MatrixDim objf_dim,
> > +                                  Real* obfj_derivs, MatrixDim derivs_dim) {
> > +  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
> > +  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
> > +  int32_cuda scores_index = i + j * scores_dim.stride;
> > +  Real K = 1.0 / (scores_dim.rows - 2.0);
> > +  Real L = scores[scores_index];
> > +  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
>
> I think you're describing an alternative way to get the _coefficients _for
> the derivative terms. But, the kernel code above already does that.
>
> On the CPU, the derivative wrt to S needs something like the following
> (NOTE: I'm ignoring peculiarities due to S being symmetric):
>
> for i=0 ... N:
> for j = 0 ... N:
> v = xvectors(i)
> w = xvectors(j)
> deriv_S += C(i,j) * (v v' + w w')
>
> Where C() is a coefficient dependent on whether or not the vectors at row
> i and j are from the same or different classes. This is what we calculated
> in the kernel above.
>
> Each v,w pair results in its own matrix. I think this makes it harder to
> deal with in a single kernel. I think the easiest thing to do is to create
> an additional kernel that works like a modified form of matrix
> multiplication. Suppose V is the matrix of xvectors and D = NumCols(V).
> Then P = V' "times" V is the serialized outer product of each row of v. For
> example, P.Row(0) = Serialized( V.Row(0) * V.Row(0)'). In other words,
> p_{i,j} = v_{i, (j / D) % D} * v_{i, j % D}.
>
> Once that is done, it should be more straightforward to calculate S_deriv
> += C(i, j) * (P.Row(i) + P.Row(j)) in parallel.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/5/files#r52834114>.
>